### PR TITLE
Fix failing float tests

### PR DESCRIPTION
**Pull request by carolinemengbq**
Float test is failing in IonHashTest.ion

Issue https://github.com/amzn/ion-hash-dotnet/issues/10

The issue was mainly caused by the conversion from double to string using Double.ToString(CultureInfo.InvariantCulture), in which some precisions were lost when dealing with many decimal digits. By using the Round-trip ("R") Format Specifier, precision is maintained. For the NaN edge case, Java and C# have differing semantics, in the test case, we are using Java's, so we map to Java's value manually.

### CODE REVIEW

File: Amazon.IonDotnet/Internals/Binary/RawBinaryWriter.cs [hunk 1]
```diff
@@ -542,7 +542,17 @@ public void WriteFloat(double value)
             {
                 _containerStack.IncreaseCurrentContainerLength(9);
                 _dataBuffer.WriteByte(TidFloatByte | 8);
-                _dataBuffer.WriteUint64(BitConverterEx.DoubleToInt64Bits(value));
+
+                if (double.IsNaN(value))
+                {
+                    // Double.NaN is different between C# and Java
+                    // For consistency, map NaN to the long value for NaN in Java
+                    _dataBuffer.WriteUint64(9221120237041090560);
```
Comment (pbcornell) [inline Amazon.IonDotnet/Internals/Binary/RawBinaryWriter.cs]: i'd prefer to see this number written in hex:  `0x7ff8000000000000L`

File: Amazon.IonDotnet/Internals/Text/IonTextRawWriter.cs [hunk 2]
```diff
@@ -228,9 +228,19 @@ public void Write(double d)
...
+
+            // Differentiate between negative zero and zero.
+            if (d == 0 && BitConverter.DoubleToInt64Bits(d) < 0)
+            {
+                str = "-0e0";
+                _writer.Write(str);
+            }
+            else
+            {
+                str = d.ToString("R");
```
Comment (pbcornell) [inline Amazon.IonDotnet/Internals/Text/IonTextRawWriter.cs]: i don't know what this is for...help me understand please?

Comment (carolinemengbq) [reply Amazon.IonDotnet/Internals/Text/IonTextRawWriter.cs]: The "R" is the round-trip format specifier, which will ensure that a numeric value that is converted to a string is parsed back into the same numeric value. The previous implementation when converting the double to string will round the decimal digits to a certain length, so some precision will be lost there.

**Comment by carolinemengbq** [comment]
> if these changes allow currently-skipped ion-dotnet tests to be re-enabled (or ion-tests files to be removed from a skip list), please re-enable them. if not, please add unit tests to assert the expected behavior
> 
> requesting review by tgregg@ as he is tracking this repository more closely

This change will enable the commented out float tests in ion_hash_tests.ion

**Comment by pbcornell** [comment]
> This change will enable the commented out float tests in ion_hash_tests.ion

That's good, but ion-dotnet needs its own testing of this behavior independent of ion-hash-dotnet.

File: Amazon.IonDotnet/Internals/Text/IonTextRawWriter.cs (hunk 2 above)
Comment (matthewh-BQ) [reply Amazon.IonDotnet/Internals/Text/IonTextRawWriter.cs]: Could we also add a comment in here to clarify this?
